### PR TITLE
implement macros for recursive closures

### PR DESCRIPTION
**Pull request by Hegdahl**

### CODE REVIEW

File: hcpl_recursion/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            }
+            pub struct Almost<F>(F);
+            impl<F: Fn(&dyn AlmostFTrait $(,$arg_ty)*) -> $ret> AlmostFTrait for Almost<F> {
+                #[inline(always)]
+                fn call(&self, $($arg_id: $arg_ty),*) -> $ret {
+                    (self.0)(self, $($arg_id),*)
+                }
+            }
+            let almost_f = Almost(|almost_f: &dyn AlmostFTrait,  $($arg_id: $arg_ty),*| -> $ret {
+                let $f = |$($arg_id: $arg_ty),*| {
```
Comment (jakobrs) [inline hcpl_recursion/src/lib.rs:17]:
```suggestion
                #[allow(unused_variables)]
                let $f = |$($arg_id: $arg_ty),*| {
```

File: hcpl_recursion/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
+#[macro_export]
+macro_rules! _recursion__let_rec {
+    ($f:ident = |$($arg_id:ident: $arg_ty:ty),*| -> $ret:ty $body:block) => {
+        let $f = {
+            pub trait AlmostFTrait {
+                fn call(&self, $($arg_id: $arg_ty),*) -> $ret;
+            }
+            pub struct Almost<F>(F);
```
Comment (jakobrs) [inline hcpl_recursion/src/lib.rs]: `pub` er unødvendig her

```suggestion
            trait AlmostFTrait {
                fn call(&self, $($arg_id: $arg_ty),*) -> $ret;
            }
            struct Almost<F>(F);
```

File: hcpl_recursion/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            }
+        };
+    }
+}
+
+#[macro_export]
+macro_rules! _recursion__let_rec_mut__impl {
+    ($f:ident = [ $($cap_id:ident: $cap_ty:ty),* $(,)? ] |$($arg_id:ident: $arg_ty:ty),*| -> $ret:ty $body:block $dol:tt) => {
+        hcpl_recursion::let_rec!(without_cap = |$($cap_id: &mut $cap_ty,)* $($arg_id: $arg_ty),*| -> $ret {
+            macro_rules! $f {
```
Comment (jakobrs) [inline hcpl_recursion/src/lib.rs]:
```suggestion
            #[allow(unused_macros)]
            macro_rules! $f {
```

**Review by jakobrs — CHANGES_REQUESTED** [review]
Bare et par ting jeg la merke til

**Review by Hegdahl — COMMENTED** [review]
Nice, accepted the changes that are not already fixed :rocket:

**Review by jakobrs — APPROVED** [review]

**Comment by jakobrs** [comment]
`$inner_args:tt` should be `$inner_args:expr`

**Pull request merged**